### PR TITLE
implement dim statement (#283)

### PR DESCRIPTION
**Pull request by strattonbrazil**
This implements the dim statement for as requested for #283. It doesn't match Roku's spec exactly as Roku's dim will leave the bottom-level arrays unexpanded but resizeable. When brs supports resizeable arrays (#530) we can come back and update this code to match the spec.

### CODE REVIEW

**Review by sjbarag — CHANGES_REQUESTED** [review]
Looking good!  Nice job jumping into the parser — this is almost perfect.  I've got two small requests that should help out with error checking and location tracking, but they shouldn't take terribly long to implement and will make a big difference.

Thanks for the thorough comments, @strattonbrazil !

File: src/parser/Parser.ts [hunk 1]
```diff
@@ -1149,6 +1153,32 @@ export class Parser {
...
+        /**
+         * Parses a `dim` statement
+         * @returns an AST representation of an `goto` statement.
+         */
+        function dimStatement() {
+            let tokens = {
+                dim: advance(),
+            };
+
+            let name = advance();
```
[the remark this answers is not in the document]
Comment (strattonbrazil) [reply src/parser/Parser.ts:1165]: Will fix.

File: src/parser/Parser.ts [hunk 2]
```diff
@@ -1149,6 +1153,32 @@ export class Parser {
...
+
+            match(Lexeme.LeftSquare);
+
+            let dimensions: Expression[] = [];
+            while (!match(Lexeme.RightSquare)) {
+                dimensions.push(expression());
+                if (check(Lexeme.Comma)) advance();
```
[the remark this answers is not in the document]
Comment (strattonbrazil) [reply src/parser/Parser.ts:1172]: Commas aren't optional (in the spec), but I need to step over them if present either before or after. I guess this current implementation does make them optional. Thoughts?

File: src/parser/Statement.ts [hunk 3]
```diff
@@ -64,6 +65,30 @@ export class Assignment extends AstNode implements Statement {
...
+        readonly tokens: {
+            dim: Token;
+        },
+        readonly name: Identifier,
+        readonly dimensions: Expr.Expression[]
+    ) {
+        super("Dim");
+    }
+
+    accept<R>(_visitor: Visitor<R>): BrsType {
```
[the remark this answers is not in the document]
Comment (strattonbrazil) [reply src/parser/Statement.ts:79]: Will fix.

File: src/parser/Statement.ts [hunk 4]
```diff
@@ -64,6 +65,30 @@ export class Assignment extends AstNode implements Statement {
...
+        return _visitor.visitDim(this);
+    }
+
+    get location() {
+        return {
+            file: this.tokens.dim.location.file,
+            start: this.tokens.dim.location.start,
+            end: this.tokens.dim.location.end,
```
[the remark this answers is not in the document]
Comment (strattonbrazil) [reply src/parser/Statement.ts:87]: I'm using the match() in the while loop. Is there a function to get the most recent token, or would I need to save the match() response (ex. while-true loop)?

File: src/parser/Parser.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjbarag) [reply src/parser/Parser.ts:1172]: You might have some luck parsing this with a single `expression()` outside the loop, then requiring a comma and an expression inside the loop!  Something like:

```typescript
let dimensions: Expression[] = [ expression() ];
while(!match(Lexeme.RightSquare)) {
    consume("Expected ',' after expression in 'dim' statement", Lexeme.Comma);
    dimensions.push(expression());
}
```

Which should work on the assumption that the pre-loop `expression()` call throws an error?  You could always `try/catch` to provide better error messaging like you do below on line 1176 😃

File: src/parser/Statement.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (sjbarag) [reply src/parser/Statement.ts:87]: You've got (at least) two options!

```typescript
let rightSquare: Token;
while (rightSquare = peek(Lexeme.RightSquare) && !match(Lexeme.RightSquare)) {
}
```

or rely on the fact that we break out of the `while` loop immediately when a `]` is encountered and use `previous()`, e.g.:

```typescript
while (!match(Lexeme.RightSquare)) {
    // …
}
let rightSquare = previous();
```

The second feels a bit more clear, but anything along those lines should be fine!

**Review by sjbarag — APPROVED** [review]
Thanks for the updates @strattonbrazil !

**Pull request merged**